tfe/resource_tfe_workspace.go

```diff
@@ -671,15 +676,54 @@ func resourceTFEWorkspaceDelete(d *schema.ResourceData, meta interface{}) error
 	id := d.Id()
 
 	log.Printf("[DEBUG] Delete workspace %s", id)
-	err := tfeClient.Workspaces.DeleteByID(ctx, id)
+
+	ws, err := tfeClient.Workspaces.ReadByID(ctx, id)
 	if err != nil {
 		if err == tfe.ErrResourceNotFound {
 			return nil
 		}
 		return fmt.Errorf(
-			"Error deleting workspace %s: %w", id, err)
+			"Error reading workspace %s: %w", id, err)
+	}
+
+	forceDelete := d.Get("force_delete").(bool)
+
+	// presence of Permissions.CanForceDelete will determine if current version of TFE supports safe deletes
+	if ws.Permissions.CanForceDelete == nil {
+		if forceDelete {
+			err = tfeClient.Workspaces.DeleteByID(ctx, id)
+		} else {
+			return fmt.Errorf(
+				"Error deleting workspace %s: This workspace must be force deleted by setting force_delete=true", id)
+		}
+	} else if *ws.Permissions.CanForceDelete {
+		if forceDelete {
+			err = tfeClient.Workspaces.DeleteByID(ctx, id)
+		} else {
+			err = tfeClient.Workspaces.SafeDeleteByID(ctx, id)
+			if err != nil {
+				if strings.Contains(err.Error(), "conflict") {
+					return fmt.Errorf(
+						"Error deleting workspace %s: %w\nTo delete this workspace without destroying the managed resources, add force_delete=true to the resource config", id, err)
+				}
+			}
+		}
+	} else {
+		if forceDelete {
+			return fmt.Errorf(
+				"Error deleting workspace %s: missing required permissions to set force_delete=true", id)
+		} else {
+			err = tfeClient.Workspaces.SafeDeleteByID(ctx, id)
+		}
 	}
```

Comment:
Disclaimer: I'm still learning Golang.

The number of nested `if` statements here is a little high, right? I think one way I think we could trim them down is to move the [error handling at the bottom](https://github.com/hashicorp/terraform-provider-tfe/pull/675/files#diff-e727ae216721eb3c1f24911e8b1ececae3ae9208f19b886eb627c76236471e13R720-R726) into a helper function. That way you could take advantage of some early returns in a few places, for example, instead of:

```go
if ws.Permissions.CanForceDelete == nil {
  // ... logic for old tfe...
} else {
  // ... logic for new tfe...
}

if err {
  // common error handling...
}
```

you could do:

```go
if ws.Permissions.CanForceDelete == nil {
  // ... logic for old tfe...
  return commonErrorHandlingHelper(err)
}

// ... logic for new tfe...
return commonErrorHandlingHelper(err)
```

Just an idea 🤙 I'm sure there are even better ways than this to trim things down you or other folks more familiar with golang than me might have

Reply:
i feel the same way about the nesting here -- i'll give this a try !